### PR TITLE
docs: component principles and createComponent readme

### PR DESCRIPTION
**Pull request by jdhuntington**
Goal: Establish a set of component design principles and document a set of architecture patterns based on those principles.

### CODE REVIEW

File: docs/src/prototypes/ComponentArchitecture/DemoTextControl.tsx
Comment (aneeshack4) [inline docs/src/prototypes/ComponentArchitecture/DemoTextControl.tsx]: Just clarifying, we're setting a pattern of the "Base" controls being super bare bones/theme agnostic/etc and the "Demo" ones are composed "Base" controls that can be themed?

File: docs/src/prototypes/ComponentArchitecture/BaseTextControl.tsx
Comment (aneeshack4) [inline docs/src/prototypes/ComponentArchitecture/BaseTextControl.tsx]: And "Base" controls can't be themed at all but can be styled if directly passed in classnames?

File: docs/src/prototypes/ComponentArchitecture/theme.ts
Comment (aneeshack4) [inline docs/src/prototypes/ComponentArchitecture/theme.ts]: This is where the new design tokens would be set right (at least their default values)?

File: packages/react-theming/README.md
Comment (aneeshack4) [inline packages/react-theming/README.md]: Would these default token/styling values would passed directly to those "Base" components with classNames?

File: packages/react-theming/README.md
Comment (aneeshack4) [inline packages/react-theming/README.md]: So the main difference between slotProps and slots is the layer at which each can be set?

File: packages/react-theming/src/compose.ts
Comment (aneeshack4) [inline packages/react-theming/src/compose.ts]: I looked at themeContext.ts where this hook is defined - so there won't be anything like the partial theme objects in fabric's theme.ts that allowed for more flexibility in terms of what the user had to define for the theme object? Instead it'll be a full theme object at every step.

File: docs/src/prototypes/ComponentArchitecture/theme.ts
Comment (aneeshack4) [inline docs/src/prototypes/ComponentArchitecture/theme.ts]: Animation will be from theme too? And it would be set here like fonts?

Comment (aneeshack4) [reply docs/src/prototypes/ComponentArchitecture/theme.ts]: Where would RTL and isInverted (like in the current scheme object in fabric) be set?

**Review by aneeshack4 — COMMENTED** [review]
Looking really good!

File: docs/src/prototypes/ComponentArchitecture/DemoTextControl.tsx
Comment (jdhuntington) [reply docs/src/prototypes/ComponentArchitecture/DemoTextControl.tsx]: Yeah, exactly correct.

File: docs/src/prototypes/ComponentArchitecture/BaseTextControl.tsx
Comment (jdhuntington) [reply docs/src/prototypes/ComponentArchitecture/BaseTextControl.tsx]: Correct

File: docs/src/prototypes/ComponentArchitecture/theme.ts
Comment (jdhuntington) [reply docs/src/prototypes/ComponentArchitecture/theme.ts]: There are still several options for where design tokens would live, but this is definitely a possible location.

File: docs/src/pages/ComponentArchitecture.mdx
Comment (khmakoto) [inline docs/src/pages/ComponentArchitecture.mdx]: nit: Extra line?

File: docs/src/pages/ComponentArchitecture.mdx
Comment (khmakoto) [inline docs/src/pages/ComponentArchitecture.mdx]:
```suggestion
of implementation options (i.e. a `accessibilityMode` prop that accepts a fixed set of string values as input).
```

File: docs/src/pages/ComponentArchitecture.mdx
Comment (khmakoto) [inline docs/src/pages/ComponentArchitecture.mdx]:
```suggestion
Further, accessibility is closely tied to a component's internal state, and they are likely to iterate concurrently.
```

File: docs/src/pages/ComponentArchitecture.mdx
Comment (khmakoto) [inline docs/src/pages/ComponentArchitecture.mdx]: Maybe?

```suggestion
pre-scaffolded with Fluent design opinions (themed components). Fluent will be opinionated but customizable.
```

File: docs/src/pages/ComponentArchitecture.mdx
Comment (khmakoto) [inline docs/src/pages/ComponentArchitecture.mdx]:
```suggestion
Themed components are the glue that ties together context, theming, and the base components. There are several potential
```

File: docs/src/pages/ComponentArchitecture.mdx
Comment (khmakoto) [inline docs/src/pages/ComponentArchitecture.mdx]:
```suggestion
the slots prop.
```

File: docs/src/pages/ComponentArchitecture.mdx
Comment (khmakoto) [inline docs/src/pages/ComponentArchitecture.mdx]:
```suggestion
## Themed components should have a default out-of-the-box theme
```

File: docs/src/pages/ComponentArchitecture.mdx
Comment (khmakoto) [inline docs/src/pages/ComponentArchitecture.mdx]:
```suggestion
The base control should leverage a platform library's function to pass all relevant native props to appropriate
```

File: docs/src/pages/ComponentArchitecture.mdx
Comment (khmakoto) [inline docs/src/pages/ComponentArchitecture.mdx]:
```suggestion
opt-in to specific versions of new components without requiring a
```

File: docs/src/prototypes/ComponentArchitecture/BasePrefixedInput.tsx
Comment (khmakoto) [inline docs/src/prototypes/ComponentArchitecture/BasePrefixedInput.tsx]: Are we not using `;` at the end of statements now?

File: docs/src/prototypes/ComponentArchitecture/DemoTextControl.tsx
Comment (khmakoto) [inline docs/src/prototypes/ComponentArchitecture/DemoTextControl.tsx]: From what I can see there's some typing enforcement missing here that I assume will be added later, right?

File: packages/react-theming/README.md
Comment (khmakoto) [inline packages/react-theming/README.md]:
```suggestion
There is no requirement around using components built with `compose`.
```

File: packages/react-theming/README.md
Comment (khmakoto) [inline packages/react-theming/README.md]:
```suggestion
pass class names to components, which will render the expected
```

File: packages/react-theming/README.md
Comment (khmakoto) [inline packages/react-theming/README.md]:
```suggestion
`slots` are defined in `createComponent` by specifying a slots argument:
```

File: packages/react-theming/README.md
Comment (khmakoto) [inline packages/react-theming/README.md]: Is it going to be `createComponent` or `compose`? I see both terms being used at different times in this PR.

File: packages/react-theming/README.md
Comment (khmakoto) [inline packages/react-theming/README.md]:
```suggestion
simply add one test per slot alongside the implementation and run
```

File: packages/react-theming/README.md
Comment (mnajdova) [inline packages/react-theming/README.md]: Aren't we going to re-calculate for different props combination? How can only the theme be key for the cache? Based on what was priviosly written:

>  A variant is defined as an additional prop on the created component which has
the ability to adjust the look and feel of the component.

it seems that we will need to include the props combination in the cache.

File: packages/react-theming/README.md
Comment (mnajdova) [inline packages/react-theming/README.md]: We discussed multiple times about replacing `slotProps` with direct props on the component. The slots prop is a great idea, but for the slotProps I still don't get the benefits we would get from it.

File: packages/react-theming/README.md
Comment (jdhuntington) [reply packages/react-theming/README.md]: Yes, this is a good point. This work needs to be done still however, these docs are just documenting the existing code.

File: packages/react-theming/README.md
Comment (jdhuntington) [reply packages/react-theming/README.md]: The main advantages are to easily replace the implementation without having to re-handle all of the props. I think we agree of having that interface at the top level, but for a base component I'm not sure it makes sense.

File: packages/react-theming/README.md
Comment (mnajdova) [reply packages/react-theming/README.md]: What existing code is that? I thought we are defining the vision of the principles and createComponent. Is this a doc for the current compose/createComponent that is part of the react-theming mechanism?

File: docs/src/pages/ComponentArchitecture.mdx
Comment (mnajdova) [inline docs/src/pages/ComponentArchitecture.mdx]: A few more points that should be mention regarding theming, based on the client's requirements we have had so far:

# Theming

The theming mechanism as part of Fluent should be flexible and provide support for as many scenarios as possible (CSS stylesheets, as well as CSSinJS solution that will not tight the users to just one CSS in JS framework).

#### A component should be able to render any theme (including tokens, styles, icons and animations)

The client should be able to reuse the components between different themes. Supporting new theme in the application should not requre building new components.

If new variants are added in the new theme, then the clients should be able to compose the existing component, by adding new design terms (via the props API).

#### Themes should cascade and merge allowing subtree themes to be updated at runtime and have the ability to adjust the incoming theme.

If some parameters/tokens are changed on the root of the application, the changes should be propagated to all nested subtrees.

#### RTL mode should be configurable for the entire application or a subtree of the application.

It is important to allow the users to specify the RTL/LTR mode for the entire application once, but if necessary they can specify it only in a subtree of the application (for example one ChatMessage)

File: docs/src/pages/ComponentArchitecture.mdx
Comment (mnajdova) [inline docs/src/pages/ComponentArchitecture.mdx]: If HoC refers to Higher order component, can we avoid it, as it implies details in implementation that we haven't decided on. I would really be against composing components and polluting the react tree.

File: docs/src/pages/ComponentArchitecture.mdx
Comment (layershifter) [inline docs/src/pages/ComponentArchitecture.mdx]:
```suggestion
- Consumers should not be penalized for code they don't use (through bundle size, perf, etc).
```

_nit_

File: docs/src/pages/ComponentArchitecture.mdx
Comment (layershifter) [inline docs/src/pages/ComponentArchitecture.mdx]:
```suggestion
```

File: docs/src/pages/ComponentArchitecture.mdx
Comment (layershifter) [inline docs/src/pages/ComponentArchitecture.mdx]:
```suggestion
Components should expect a prop called `slots`, which is a dictionary of element names mapped
```

Another one, should be configured under components or during compose?

File: docs/src/pages/ComponentArchitecture.mdx
Comment (layershifter) [inline docs/src/pages/ComponentArchitecture.mdx]:
```suggestion
Hooks should be created to manage internal state of components. For sets
```

File: docs/src/pages/ComponentArchitecture.mdx
Comment (layershifter) [inline docs/src/pages/ComponentArchitecture.mdx]: I suggest to delete the whole paragraph as we don't have any agreement what is "base components".

File: docs/src/pages/ComponentArchitecture.mdx
Comment (layershifter) [inline docs/src/pages/ComponentArchitecture.mdx]: Let's also delete this as it's opinionated about CSS/CSS-in-JS usage.

File: docs/src/pages/ComponentArchitecture.mdx
Comment (layershifter) [inline docs/src/pages/ComponentArchitecture.mdx]: #### Do not force to use any CSS or CSS-in-JS implementation

Applications which load their own CSS in other means should not pay the penalty of loading a runtime CSS-in-JS package, especially if the class names are known at time of authoring.

#### A component should be able to render any theme (including tokens, styles, and icons) or any CSS

???

#### Theming is first-class citizen

A proposed CSS-in-JS solution should allow themes to cascade and merge allowing subtree themes to be updated at runtime and have the ability to adjust the incoming theme.

#### RTL mode should be configurable for the entire application or a subtree of the application

```tsx
// 👍, `Provider` is an abstract name
<Provider direction='rtl' />
// 👎 
<>
  <Button direction='rtl' />
  <Status direction='rtl' />
</>
```

File: docs/src/pages/ComponentArchitecture.mdx
Comment (layershifter) [inline docs/src/pages/ComponentArchitecture.mdx:20]: #### Components should be composable

Composition should allow to apply changes to accessibility, styles or state of components.

```tsx
<Image /> {/* <img src="" /> */}
<RemovedImage /> {/* <img alt="This image was removed" /> */}
```

#### There should be one JSX element in the tree per rendered component

```tsx
// 👍 this is fine
<AvatarImage />
// 👎 bloats React tree
<AvatarImage>
  <Image />
</AvatarImage>
```

File: docs/src/pages/ComponentArchitecture.mdx
Comment (layershifter) [inline docs/src/pages/ComponentArchitecture.mdx]: Let's exclude the idea of "Base Components" at all and let's try to generalize principles without any implementation details as it creates borders and edges for other proposals.

File: packages/react-theming/README.md
Comment (jdhuntington) [reply packages/react-theming/README.md]: No, components used without compose would have no access to tokens; it would be the caller's responsibility to provide classes.

File: packages/react-theming/README.md
Comment (jdhuntington) [reply packages/react-theming/README.md]: They can both be accessed most places. The difference is what goes in each (props for a component, or a component reference itself). The value of keeping them separate is to make replacing a base's component with a themed component an easy process.

File: packages/react-theming/src/compose.ts
Comment (jdhuntington) [reply packages/react-theming/src/compose.ts]: I believe that's still TBD, but it would be nice if compose could assume a full theme object at each step.

File: docs/src/prototypes/ComponentArchitecture/theme.ts
Comment (jdhuntington) [reply docs/src/prototypes/ComponentArchitecture/theme.ts]: Yes, all of those things should be here. Most of this is stubbed out and waiting for a final answer. 😄

File: docs/src/prototypes/ComponentArchitecture/BasePrefixedInput.tsx
Comment (jdhuntington) [reply docs/src/prototypes/ComponentArchitecture/BasePrefixedInput.tsx]: No. Most all of this repo is configured via prettier to omit semicolons.

File: docs/src/prototypes/ComponentArchitecture/DemoTextControl.tsx
Comment (jdhuntington) [reply docs/src/prototypes/ComponentArchitecture/DemoTextControl.tsx]: Exactly.

File: packages/react-theming/README.md
Comment (jdhuntington) [reply packages/react-theming/README.md]: Ahh, good call. I meant to keep this consistent. It's `compose` now, but we are going to change it as several FP libraries already have a `compose` impl that differs.

File: packages/react-theming/README.md
Comment (jdhuntington) [reply packages/react-theming/README.md]: Yes, sorry for muddying the waters. I will remove the README changes from this PR so it's more focused.

File: docs/src/pages/ComponentArchitecture.mdx
Comment (jdhuntington) [reply docs/src/pages/ComponentArchitecture.mdx]: Happy to discuss the pros/cons, but deciding on base component structure (including whether or not they even exist) is the main point of this doc. I think the borders and edges are the value proposition, as it makes theming/composition discussions clear and decoupled from component implementation discussions.

File: docs/src/pages/ComponentArchitecture.mdx
Comment (jdhuntington) [reply docs/src/pages/ComponentArchitecture.mdx]: Added the last one, though I would like to discuss the first 3 in another PR.

File: docs/src/pages/ComponentArchitecture.mdx
Comment (jdhuntington) [reply docs/src/pages/ComponentArchitecture.mdx]: I think this is true of base components, and compose/createComponents will provide that opinion.

File: docs/src/pages/ComponentArchitecture.mdx
Comment (levithomason) [reply docs/src/pages/ComponentArchitecture.mdx]: Offline agreement: split this PR into arch requirements only (the "why"), then, as code is committed in separate PRs to fulfill those requirements, we can document "how" it is accomplished in Fluent UI.

This way, we're committing the description along with the real-world working thing.

File: docs/src/pages/ComponentArchitecture.mdx
Comment (levithomason) [reply docs/src/pages/ComponentArchitecture.mdx:20]: Suggested update:

```diff
-Composition should allow to apply changes to accessibility, styles or state of components.
+Composition should allow to change the accessibility, styles, state, or slots of a component.
```

File: docs/src/pages/ComponentArchitecture.mdx
Comment (levithomason) [reply docs/src/pages/ComponentArchitecture.mdx]: The first point on CSS-in-JS solutions probably falls under our more broad requirement/philosophy:

> ### The consumer should be able to override, extend, or replace any opinions of the library.
> State ...
> styling ...
> A11y ...
> Slots ..
> Styling approach (css, css-in-js, etc) ...
> css-in-js renderer ...

Something like this.  It may be done through composition or flexible components but the need to change the essence of the component is what we're targeting.

File: docs/src/pages/ComponentArchitecture.mdx
Comment (levithomason) [reply docs/src/pages/ComponentArchitecture.mdx]: Let's leave this as a messy list of "all the requirements" for now.  As we split Fluent UI into more clear layers (base, fluent, `<product>`) we can split these requirements and arch points into "base" and "fluent design" or appropriate.

File: docs/src/pages/ComponentArchitecture.mdx
Comment (levithomason) [inline docs/src/pages/ComponentArchitecture.mdx]: Suggestion for stating this void of implementation details:

>Consumers should be able to import one component and have an appropriately minimal bundle size

File: docs/src/pages/ComponentArchitecture.mdx
Comment (levithomason) [reply docs/src/pages/ComponentArchitecture.mdx]: Suggest this the same requirement as above, see comment 
https://github.com/microsoft/fluent-ui-react/pull/2255/files#r370419830

File: docs/src/pages/ComponentArchitecture.mdx
Comment (levithomason) [inline docs/src/pages/ComponentArchitecture.mdx]: This point `#### Themed components pass ...` and explainer is great.  Let's keep but strip the details about theme component etc.  Keep it to how we handle known/safe props.

File: packages/react-theming/README.md
Comment (mnajdova) [reply packages/react-theming/README.md]: I see now, I thought everything is part of the component principles write up. Yep, it would be best if we can extract it. Thanks!

File: docs/src/pages/ComponentArchitecture.mdx
Comment (mnajdova) [inline docs/src/pages/ComponentArchitecture.mdx]: Can you simplify the second sentence, specifically what iterate concurrently means. It's a bit confusing.

File: docs/src/pages/ComponentArchitecture.mdx
Comment (mnajdova) [inline docs/src/pages/ComponentArchitecture.mdx:39]: This may be a bit tricky. Take a look on this example:
```tsx
<Input autoComplete />
```
```html
<div>
  <input type="text" />
</div>
```
The autoComplete in this case should not be propagated to the root, but to the input slot.

File: docs/src/pages/ComponentArchitecture.mdx
Comment (mnajdova) [inline docs/src/pages/ComponentArchitecture.mdx]: We have also requirement for disabling animations, which can be configured for the entire application or a subtree of the application.

This is a real use-case, where we are disabling animations for the whole applications, but the loaders are still animated.

**Review by mnajdova — APPROVED** [review]
Some final comments, other that that looks good 👍

**Review by layershifter — APPROVED** [review]
Please check comments from @mnajdova before merging 👍

File: docs/src/pages/ComponentArchitecture.mdx
Comment (jdhuntington) [reply docs/src/pages/ComponentArchitecture.mdx:39]: Agreed. Based on this principle, it would be up to the component to redirect that prop to the `input`.

Comment (jdhuntington) [reply docs/src/pages/ComponentArchitecture.mdx:39]: If it's questionable, I'm happy to remove this one. I think this makes sense as a foundation though and we can special case when necessary.

File: docs/src/pages/ComponentArchitecture.mdx
Comment (jdhuntington) [reply docs/src/pages/ComponentArchitecture.mdx]: Actually I think it's more clear if it's omitted entirely. Thanks.

**Pull request merged**
